src/types/chat.rs

```diff
@@ -0,0 +1,445 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::error::{Error, FromError};
+use std::io;
+
+use rustc_serialize::json;
+use rustc_serialize::json::{Json, ToJson};
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ChatJsonError {
+    MalformedJson(json::ParserError),
+    IoError(io::Error),
+    NotAnObject,
+    InvalidFieldType,
+    UnknownField(String),
+    InvalidColor(String),
+    InvalidClickEvent,
+    InvalidHoverEvent
+}
+
+impl FromError<io::Error> for ChatJsonError {
+    fn from_error(err: io::Error) -> ChatJsonError {
+        ChatJsonError::IoError(err)
+    }
+}
+
+impl FromError<json::ParserError> for ChatJsonError {
+    fn from_error(err: json::ParserError) -> ChatJsonError {
+        if let json::ParserError::IoError(e) = err {
+            ChatJsonError::IoError(e)
+        } else {
+            ChatJsonError::MalformedJson(err)
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ChatJson {
+    pub msg: Message,
+    pub extra: Option<Vec<Json>>,
+    pub color: Option<Color>,
+    pub formats: BTreeSet<Format>,
+    pub click_event: Option<ClickEvent>,
+    pub hover_event: Option<HoverEvent>,
+    pub insertion: Option<String>
+}
+
+impl ChatJson {
+    pub fn msg(msg: String) -> ChatJson {
+        ChatJson {
+            msg: Message::PlainText(msg), extra: None, color: None, formats: BTreeSet::new(),
+            click_event: None, hover_event: None, insertion: None
+        }
+    }
+
+    pub fn from_json(src: &mut io::Read) -> Result<ChatJson, ChatJsonError> {
```

Comment:
This should be called `from_reader` or similar. An actual `from_json` might also be helpful.